Create ReleasePlanning.md

mkArtakMSFT opened the pull request:
We've changed the way we do release planning this time around so I want to write it down to remember in the future and polish this process as we learn of new ways for doing so.

@dotnet/aspnet-blazor-eng, let's make sure we all are on the same page regarding this process. This writeup will help with that. If you see anything which should be adjusted, please share your notes in comments.

@rafikiassumani-msft, @adityamandaleeka I'm not sure whether this will be helpful for you too or not. If you would like to adopt this model, feel free to leave comments. Alternatively, I'm happy to rename this file to reflect that this is the process only my team uses to avoid confusion.

mkArtakMSFT committed:
Create ReleasePlanning.md

TanayParikh commented on docs/ReleasePlanning.md:
```diff
@@ -0,0 +1,41 @@
+Throughout the year we add issues to the `Backlog` milestone as is pointed out in our [Triage Process](./docs/TriageProcess.md).
+We review all the issues in that milestone once a year, after the work on an upcoming major release is complete.
+Given the large number of issue, it takes multiple sessions for teams to review and identify candidates for consideration for the next major release.
```
```suggestion
Given the large number of issues, it takes multiple sessions for teams to review and identify candidates for consideration for the next major release.
```

TanayParikh commented on docs/ReleasePlanning.md:
```diff
@@ -0,0 +1,41 @@
+This document details the process we use for identifying candidate issues for the next release.
+
+## Phases
+The process for identifying candidates for the next major release consistes of multiple phases, each one filtering out issues by leaving the min backlog or closing them.
```
```suggestion
The process for identifying candidates for the next major release consists of multiple phases. In each phase, we filter issues out of the release by either moving them to the `backlog`, or closing the issue.
```

TanayParikh commented on docs/ReleasePlanning.md:
```diff
@@ -0,0 +1,41 @@
+- Define the cut line
+
+### Filtering
+At this stage all the issues are distributed to engineers by feature areas.
+The team agrees on some number of issues for each engineer to bring in to the next meeting with individual priority labels assigned  - fl-p1, fl-p2, fl-p3, where `fl` are the first letters of their first and last name.
+
+Each engineer then goes through every single issue in the set they've been assigned to and labels those with individual priority labels.
+All the issues which the engineer thinks is lower than p3 - remains on the backlog. We also agree to approximately balance the distribution of the 3 priroity labels on the issues that will be brough back by each engineer, so that it forces real prioritization excercise.
+The issues which engineers think are good candidates and fit in the above listed requirements are moved to the `.NET V Planning` milestone, where `V` is the updcoming version number (7 at the time of this writing) - `.NET 7 Planning`.
```
```suggestion
At this stage all the issues are distributed to engineers by feature areas. Each engineer reviews all the issues within their feature area, and returns to the next meeting with individual priority labels assigned - fl-p1, fl-p2, fl-p3, where `fl` are their initials.

All the issues which the engineer believes are lower than `Priority-3` - remain in the backlog. We also agree to approximately balance the distribution of the 3 priority labels on the issues that will be brought back by each engineer, so that it forces real prioritization exercise.
The issues which engineers think are good candidates and fit in the above listed requirements are moved to the `.NET V Planning` milestone, where `V` is the updcoming version number (7 at the time of this writing) - `.NET 7 Planning`.
```

TanayParikh commented on docs/ReleasePlanning.md:
```diff
@@ -0,0 +1,41 @@
+This will be used later during the planning process.
+
+### Team Review & Priority adjustment
+Now, that all the issues are in the `.NET V planning` milestone, the team reviewes each issues one at a time starting from the highest priority ones (P1).
```
```suggestion
Now, that all the issues are in the `.NET V planning` milestone, the team reviews each issue one at a time starting from the highest priority ones (P1).
```

Should this be P0?

TanayParikh commented on docs/ReleasePlanning.md:
```diff
@@ -0,0 +1,41 @@
+### Team Review & Priority adjustment
+Now, that all the issues are in the `.NET V planning` milestone, the team reviewes each issues one at a time starting from the highest priority ones (P1).
+We discuss the issues and agree on the priority at this point. Sometimes we make adjustments to the suggested individual priorities. After discussing each issue the `Priority: X` label is applied to each issue.
+Each `Priority: 1` issue is then moved to the project board, which will be used by each team for tracking the work for the upcoming release throughout the year. The project board has a `Triage` column and that's where all these issues land at this point.
+At this point we bring to the board only the top prioritiy issues.
```
```suggestion
Each `Priority: 1` issue is then moved to the project board, which will be used by each team for tracking the work for the upcoming release throughout the year. The issues start off in the `Triage` column. At this point we bring only the top priority issues to the board.
```

TanayParikh commented on docs/ReleasePlanning.md:
```diff
@@ -0,0 +1,41 @@
+At this point we bring to the board only the top prioritiy issues.
+
+### Capacity Planning
+We usually reserve only 50% of the team capacity for this work. The reason is that we will be getting a lot of incomingfeedback throughout the year and we need to allocate time for handling this feedback throughout the year.
```
```suggestion
We usually reserve only 50% of the team capacity for this work. The reason is that we will be getting a lot of incoming feedback throughout the year and we need to allocate time for handling this feedback throughout the year.
```

TanayParikh commented on docs/ReleasePlanning.md:
```diff
@@ -0,0 +1,41 @@
+So we calculate the capacity of the team in weeks for the upcoming year and use half of the final number later in this process.
+
+### Define the cut line
+At this point we have all the candidate issues that we think is worth considering for the upcoming release. This number is quite large, so the teams usually won't have enough capacity to handle all this.
```
```suggestion
At this point we have all the candidate issues that we think are worth considering for the upcoming release. This number is quite large, so the teams usually won't have enough capacity to handle all this.
```

rafikiassumani-msft commented:
@mkArtakMSFT some of the steps are similar to what we started doing in `web-frameworks` team.  We are planning to do lots of e2e spikes for most of the mvc bugs to create a reliable estimate (costing) and engineering approaches to solve the issues. Prioritization is something that will be interesting to discuss as a group.

mkArtakMSFT committed:
Update docs/ReleasePlanning.md

Co-authored-by: Tanay Parikh <TanayParikh@users.noreply.github.com>

mkArtakMSFT committed:
Update docs/ReleasePlanning.md

Co-authored-by: Tanay Parikh <TanayParikh@users.noreply.github.com>

mkArtakMSFT committed:
Update docs/ReleasePlanning.md

Co-authored-by: Tanay Parikh <TanayParikh@users.noreply.github.com>

mkArtakMSFT committed:
Update docs/ReleasePlanning.md

Co-authored-by: Tanay Parikh <TanayParikh@users.noreply.github.com>

mkArtakMSFT committed:
Update docs/ReleasePlanning.md

Co-authored-by: Tanay Parikh <TanayParikh@users.noreply.github.com>

mkArtakMSFT committed:
Update docs/ReleasePlanning.md

Co-authored-by: Tanay Parikh <TanayParikh@users.noreply.github.com>

Pilchie commented:
I think one thing this write-up assumes is team members who are already quite familiar with the areas, and that typically the person doing the costing will be the one doing the work.  That won't always be the case, so it can be interesting to try things like planning poker to set the costs, so that you can different perspectives on what will be hard vs easy.

One thing I really like is doing some federation of the initial prioritization process. We tend towards consensus decision making, and it can really drag on.  Hopefully this federation can help make this process faster.

This looks like it doesn't include a step for closing issues in the backlog that aren't selected.  Is that intentional?  There are definitely two schools of thought on it, and it's something we should be consistent about among different teams of this repo.

mkArtakMSFT committed:
Update docs/ReleasePlanning.md

Co-authored-by: Tanay Parikh <TanayParikh@users.noreply.github.com>

mkArtakMSFT commented:
> I think one thing this write-up assumes is team members who are already quite familiar with the areas, and that typically the person doing the costing will be the one doing the work. That won't always be the case, so it can be interesting to try things like planning poker to set the costs, so that you can different perspectives on what will be hard vs easy.
> One thing I really like is doing some federation of the initial prioritization process. We tend towards consensus decision making, and it can really drag on. Hopefully this federation can help make this process faster.

Estimates are never precise, especially those defined a while ago. Hence, this process sets the estimates for short-term use. Specifically, those estimates will be used for drawing approximate cut-line for work which will go into the upcoming release (.NET 7 in this case). As for who will be working on those, given that we take estimates from different people into this process, the approximation works out well as it averages out in the sum.
Having said this, costing exercises are useful for short-term use. Closes thing for that would be our milestone planning process, where we don't rely on costing. But that's a different topic.


> This looks like it doesn't include a step for closing issues in the backlog that aren't selected. Is that intentional? There are definitely two schools of thought on it, and it's something we should be consistent about among different teams of this repo.

As a result of this process, issues which remain on the Backlog can be processed separately afterwards and those which are left in Backlog for more than 2 releases can be closed. I think that is a separate process, but I'm happy to include that as part of this writeup. Let me know what your preferences are on this.

Pilchie reviewed (approved):
I think it would be good to define what expectations are around the backlog.  On costing, I think things to watch out for are:

1. Relying on out of date information from costs - as long as we know that these aren't durable, it's probably fine.
2. Summarizing knowledge in a form that isn't durable. As people are determining costs, whatever they can do to capture their current understanding of the work and potential fix is valuable. Otherwise people end up having to rebuild a bunch of that context later when they start fixing the issue.

mkArtakMSFT committed:
Update ReleasePlanning.md

mkArtakMSFT committed:
Update ReleasePlanning.md

mkArtakMSFT commented:
Thank you, @Pilchie. I've updated the writeup to include some more details addressing your suggestions. As for the issues in the backlog, I plan to also update the Triage Process document, and include a writeup about Backlog cleanup in it.

mkArtakMSFT committed:
Updated the triage process to reflect the updated process for .NET 7

- Updated the process to account for .NET 7 planning process
- Added a section reflecting the Backlog cleanup process

mkArtakMSFT commented:
🆙 📅
